### PR TITLE
Fix updating choices in `gr.Dropdown` and updates related to other components

### PR DESCRIPTION
**Pull request by abidlabs**
Instead of sending the update dictionary directly to the frontend, we need to update it with the final values of the attributes from the Blocks instance. This approach should work generally, even for custom components.

Test with:

```py
import gradio as gr

SITES_SUPPORTED = ['Bedford', 'Luton']
DATES_SUPPORTED = {'Bedford': ['JAN', 'FEB', 'MAR'], 'Luton': ['SEP', 'OCT', 'NOV']}

def update_dropdown_choices(site_selected):
    selected_dates = DATES_SUPPORTED[site_selected]
    print(selected_dates)
    return gr.Dropdown(choices=selected_dates, value=selected_dates[0], visible=True)

with gr.Blocks() as demo:
    sites_dropdown = gr.Dropdown(choices=SITES_SUPPORTED, label="Sites", interactive=True)
    dates_availability = gr.Dropdown(choices=[], label="Dates", interactive=True)
    sites_dropdown.select(update_dropdown_choices, [sites_dropdown], [dates_availability])

demo.launch()
```

Fixes: #6304 

Will add tests but otherwise ready for review:
* I'll add some python unit tests for this method in `blocks.py`
* Will add a functional test because this behavior has broken several times in the past in different ways

### CODE REVIEW

**Comment by abidlabs** [comment]
Added unit tests and a functional test -- should be good for review.

**Comment by abidlabs** [comment]
The failing functional test is also happening on `main`. Started an [internal discussion](https://huggingface.slack.com/archives/C02QNREU24S/p1699309796664009)

File: gradio/blocks.py [hunk 1]
```diff
@@ -382,13 +382,16 @@ def postprocess_update_dict(
...
     """
-    update_dict = {k: update_dict[k] for k in update_dict if hasattr(block, k)}
-    if update_dict.get("value") is components._Keywords.NO_VALUE:
-        update_dict.pop("value")
-    elif "value" in update_dict and postprocess:
-        update_dict["value"] = block.postprocess(update_dict["value"])
-        if isinstance(update_dict["value"], (GradioModel, GradioRootModel)):
-            update_dict["value"] = update_dict["value"].model_dump()
+    value = update_dict.pop("value", components._Keywords.NO_VALUE)
+    print("value", value)
```
Comment (dawoodkhan82) [inline gradio/blocks.py:386]: print statement

**Review by dawoodkhan82 — APPROVED** [review]
lgtm!

File: gradio/blocks.py (hunk 1 above)
Comment (abidlabs) [inline gradio/blocks.py:386]:
```suggestion
```

**Comment by abidlabs** [comment]
Thanks for the review @dawoodkhan82!

**Pull request merged**